### ISSUE TITLE
Regent: legion_get_current_time_in_micros not safe to call from replicated task

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1209]
Someone recently pointed out to me that `legion_get_current_time_in_micros` and friends are not safe to call inside of a replicated task because the different shards may record the time at different moments. I believe that users you want to do timing inside of replicated tasks instead need to use `legion_issue_timing_op_microseconds`, which is guaranteed by the runtime to return the same future value on all shards. (The latter also returns a future, which means you'll need to wrap it in `__future`.)

I can make this change easily enough, but I wanted to post an issue so that other Regent users are aware, since using `legion_get_current_time_in_micros` for timing is pretty widespread.

**Comment by elliottslaughter** [comment]
Some clarifications:

If you have code that looks like:

```
regentlib.c.legion_get_current_time_in_micros()
```

Under the proposed change, this will need to modified to:

```
__future(regentlib.c.legion_issue_timing_op_microseconds(__runtime(), __context()))
```

There may still be (rare) cases where users occasionally need to get the actual timing on a specific shard, and are aware that this time may be different from other shards (e.g., such as when trying to compute the amount of skew between shards). In this case, the following would be used to work around the restriction, allowing you to call `legion_get_current_time_in_micros` inside a replicable task:

```
terra get_time_unsafe()
  return regentlib.c.legion_get_current_time_in_micros()
end
get_time_unsafe.replicable = true
```

It is very important to keep in mind that this will produce **different** results on different shards, so you should not rely on it for anything that may influence control flow.

No changes to the C or C++ APIs are proposed in this issue. The only change that is anticipated is what code Regent considers to be permissible inside of replicable tasks.

**Comment by lightsighter** [comment]
FWIW, there is a workaround for this. If you call `legion_issue_timing_op_*` and then wait on the resulting future, that should give you the same result across all the shards. Conceivably, we could also introduce new calls into the runtime that would do blocking immediate timing measurements across the shards. This is not something that would be terribly hard to implement since we already have all the infrastructure for it.

**Comment by mariodirenzo** [comment]
Is it possible to have in Regent a single interphase that automatically detects whether it is called in a replicated context or in a simple task and utilize the appropriate method to get the timing?
I am thinking about something like
```
terra get_wall_time_in_micros()
   if [Called from a replicable task] then
      return __future(int64, regentlib.c.legion_issue_timing_op_microseconds(__runtime(), __context()))
   else
      return regentlib.c.legion_get_current_time_in_micros()
   end
end
```

**Comment by elliottslaughter** [comment]
I could make a function like this and expose it through Regent. It will need to block when you're in a replicated task. As long as you're ok with that, I can do ahead and do this.

**Comment by mariodirenzo** [comment]
Could it return the Future and block just in case the value for the timing is utilized within the replicated task (which seems unlikely)?

**Comment by elliottslaughter** [comment]
Backing up a bit, your pseudocode above doesn't work because it has conflicting return types (`int64` vs `future(int64`). We need to pick one or the other.

 * If we pick `int64`, then we must block in the replicable case. That means it will block at the point where you call the timing function, not at the point where you use it. But I can do this in a way that can be used everywhere (replicable, inner, leaf tasks should all work). Optimal efficiency is preserved in the leaf case because we don't use any unnecessary futures there.
 * If we pick `future(int64)`, then we're introducing a potentially blocking call everywhere (leaf and non-leaf tasks alike). The runtime may be smart enough to realize that the future is already filled, but I can't promise that 100%. It also requires an additional trip through the runtime to go through the future path even in the leaf case where we don't need it. But it does allow blocking in the replicable case to be deferred to the point where the timing result is actually used.

**Comment by elliottslaughter** [comment]
I guess a follow-up question to @lightsighter: if I make a future from a concrete value, can you promise that retrieving the value will never block? That would reduce the cost of timing inside leaf tasks (though not to zero, since we're still going through some future creation and handling).

**Comment by lightsighter** [comment]
> I guess a follow-up question to @lightsighter: if I make a future from a concrete value, can you promise that retrieving the value will never block?

Yes, if it's a future that you made then I can promise you that the runtime will never block waiting on it to retrieve the value inside the future.

**Comment by elliottslaughter** [comment]
Ok, there is an implementation here that you can now try out in master and control replication. Let me know if this works for you:

https://gitlab.com/StanfordLegion/legion/-/blob/master/language/tests/regent/run_pass/timing.rg
